pkg/gpu/probe.go

```diff
@@ -0,0 +1,204 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"fmt"
+	"regexp"
+
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/rlimit"
+
+	"github.com/DataDog/datadog-agent/comp/updater/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/uprobes"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// TODO: Set a minimum kernel version
+var minimumKernelVersion = kernel.VersionCode(5, 0, 0)
```

Comment:
why do we actually need a ringbuf here? what is the benefit?

Reply:
I think event ordering is important here. The program should always see `cudaFree` after `cudaMalloc`. This is not guaranteed by perf buffers.

Secondly, perf buffers would allocate too much memory. From the current size of the ringbuffer it seems that the buffer size does not have to be large for this functionality, so ringbuffers are more memory efficient.

Reply:
Indeed, I chose ring buffers from the beginning for the guaranteed ordering (didn't know about the memory usage so that's nice too 😀), with the idea that it's easier to change back and support older kernels once we have everything working, instead of doing everything at the same time.